src/streambundle.ts

```diff
@@ -0,0 +1,189 @@
+    this.selfContext = 'vessels.' + selfId
+    this.buses = {}
+    this.allPathsBus = new Bacon.Bus()
+    this.selfBuses = {}
+    this.selfAllPathsBus = new Bacon.Bus()
+    this.selfStreams = {}
+    this.selfAllPathsStream = new Bacon.Bus()
+    this.keys = new Bacon.Bus()
+    this.availableSelfPaths = {}
+    this.app = app
+    this.metaBus = new Bacon.Bus()
+    this.selfMetaBus = new Bacon.Bus()
+  }
+
+  pushDelta(delta: Delta) {
+    const processItems = (
+      update: Update,
+      items: (Meta | PathValue)[],
+      isMeta: boolean
+    ) => {
+      if (items) {
+        items.forEach((pathValue) => {
+          this.push(pathValue.path, {
+            path: pathValue.path,
+            value: pathValue.value,
+            context: delta.context!, // TSTODO: make optional/required match
+            source: update.source,
+            $source: update.$source!, // TSTODO: make optional/required match
+            timestamp: update.timestamp!, // TSTODO: make optional/required match
+            isMeta: isMeta
+          })
+        })
+      }
+    }
+
+    try {
+      if (delta.updates) {
+        delta.updates.forEach((update) => {
+          if ('meta' in update) {
+            processItems(update, update.meta, true)
+          }
+          if ('values' in update) {
+            processItems(update, update.values, false)
+          }
+        }, this)
+      }
+    } catch (e) {
+      console.error(e)
+    }
+  }
+
+  push(path: Path, normalizedDelta: NormalizedDelta) {
+    const { isMeta } = normalizedDelta
+    const isSelf = normalizedDelta.context === this.selfContext
+    if (isMeta) {
+      const metaDelta = normalizedDelta as NormalizedDelta<true>
+      this.metaBus.push(metaDelta)
+      if (isSelf) {
+        this.selfMetaBus.push(metaDelta)
+      }
+    }
+    if (!this.availableSelfPaths[path]) {
+      this.availableSelfPaths[path] = true
+    }
+    this.getBus().push(normalizedDelta)
+    this.getBus(path).push(normalizedDelta)
+    if (isSelf) {
+      this.getSelfBus().push(normalizedDelta)
+      this.getSelfBus(path).push(normalizedDelta)
+      if (!isMeta) {
+        this.getSelfStream().push(normalizedDelta.value)
+        this.getSelfStream(path).push(normalizedDelta.value)
+      }
+    }
+  }
+
+  getMetaBus() {
+    return this.metaBus
+  }
+
+  getSelfMetaBus() {
+    return this.selfMetaBus
+  }
+
+  getBus(path?: Path) {
+    if (path !== undefined) {
+      let result = this.buses[path]
+      if (!result) {
+        result = this.buses[path] = new Bacon.Bus()
+        this.keys.push(path)
+      }
+      return result
+    } else {
+      return this.allPathsBus
+    }
+  }
+
+  getSelfStream(path?: Path) {
+    if (path !== undefined) {
+      let result = this.selfStreams[path]
+      if (!result) {
+        result = this.selfStreams[path] = new Bacon.Bus()
+      }
+      return result
+    } else {
+      return this.selfAllPathsStream
+    }
+  }
+
+  getSelfBus(path?: Path) {
```

Comment:
Should the signature require a defined value? In real life there's no point in calling these without a Path.

Reply:
It looks like it returns `selfAllPathsBus` if a `path` is `undefined`. So that might be a breaking change to require it. We could add an overload to deprecate the undefined version, but not sure how prevalent it is in the wild. 

A quick GitHub search shows [at least one plugin using it](https://github.com/chacal/signalk-stash/blob/a12230ab3527f3a579c5ae3f8c8df2af65531205/plugin/src/index.ts#L208).

Reply:
Right. I wonder however came up with that 😁